lib/wasi/src/state/func_env.rs

```diff
@@ -0,0 +1,213 @@
+                wasmer::Value::I32(a) => a as u64,
+                wasmer::Value::I64(a) => a as u64,
+                _ => DEFAULT_STACK_SIZE,
+            }
+        } else {
+            DEFAULT_STACK_SIZE
+        };
+        self.data_mut(store).stack_base = stack_base;
+
+        Ok(())
+    }
+
+    /// Like `import_object` but containing all the WASI versions detected in
+    /// the module.
+    pub fn import_object_for_all_wasi_versions(
+        &self,
+        store: &mut impl AsStoreMut,
+        module: &Module,
+    ) -> Result<Imports, WasiError> {
+        let wasi_versions =
+            get_wasi_versions(module, false).ok_or(WasiError::UnknownWasiVersion)?;
+
+        let mut resolver = Imports::new();
+        for version in wasi_versions.iter() {
+            let new_import_object =
+                crate::generate_import_object_from_env(store, &self.env, *version);
+            for ((n, m), e) in new_import_object.into_iter() {
+                resolver.define(&n, &m, e);
+            }
+        }
+
+        Ok(resolver)
+    }
+
+    pub fn cleanup(&self, store: &mut Store, exit_code: Option<ExitCode>) {
+        trace!(
+            "wasi[{}:{}]::cleanup - destroying local thread variables",
+            self.data(store).pid(),
+            self.data(store).tid()
+        );
+
+        // Destroy all the local thread variables that were allocated for this thread
+        let to_local_destroy = {
+            let thread_id = self.data(store).thread.tid();
+            let mut to_local_destroy = Vec::new();
+            let mut inner = self.data(store).process.write();
+            for ((thread, key), val) in inner.thread_local.iter() {
+                if *thread == thread_id {
+                    if let Some(user_data) = inner.thread_local_user_data.get(key) {
+                        to_local_destroy.push((*user_data, *val))
+                    }
+                }
+            }
+            inner.thread_local.retain(|(t, _), _| *t != thread_id);
+            to_local_destroy
+        };
+        if to_local_destroy.len() > 0 {
+            if let Some(thread_local_destroy) = self
+                .data(store)
+                .inner()
+                .thread_local_destroy
+                .as_ref()
+                .map(|a| a.clone())
+            {
+                for (user_data, val) in to_local_destroy {
+                    let user_data_low: u32 = (user_data & 0xFFFFFFFF) as u32;
+                    let user_data_high: u32 = (user_data >> 32) as u32;
+
+                    let val_low: u32 = (val & 0xFFFFFFFF) as u32;
+                    let val_high: u32 = (val >> 32) as u32;
+
+                    let _ = thread_local_destroy.call(
+                        store,
+                        user_data_low as i32,
+                        user_data_high as i32,
+                        val_low as i32,
+                        val_high as i32,
+                    );
+                }
+            }
+        }
```

Comment:
I think this function logic should live on the `WasiEnv` itself, since it's mainly operating with it

Reply:
The problem is that this cleanup needs to call back into the instance.
The rest of the code just collects the necessary arguments for the call. 

Doesn't make much sense to me to move it out.

Reply:
Really needs to be on the FunctionEnv.